### ISSUE TITLE
Provide an option to skip building grammars or provide paths to prebuilt ones

### ISSUE THREAD

**Issue by lmartinez-mirror** [afnanenayet/diffsitter#167]
**Is your feature request related to a problem?**

Not related to a problem.

**Context**

I'm working on packaging this for the AUR as `diffsitter` as opposed to `diffsitter-bin`, which provides a prebuilt binary. Unfortunately I can't do this as cargo insists on building the grammars listed as submodules. The issue is that I've already packaged most of these grammars separately (see [here](https://aur.archlinux.org/packages/?O=0&K=tree-sitter)), all of which compile from source.

**Describe the solution you'd like**
I'd like to see a way to either bypass compiling the grammars or provide a path to a prebuilt one.

**Describe alternatives you've considered**
As a workaround, I've tried to patch `Cargo.toml` such that it would not build the grammars, but it hasn't worked.

**Comment by afnanenayet** [comment]
Yeah the build process isn't ideal, I agree. When I started this project not all of the grammars had Rust bindings available but I think a lot of them (if not all of them) do now, so we should be able to specify them as dependencies in the Cargo manifest.

I was literally just about to open an issue about the Windows CI being flaky (seems to fail on the linkage step from time to time), and I wonder if ditching the manual compilation step would fix that too.

**Comment by afnanenayet** [comment]
Just to clarify, it's not Cargo that's trying to build the submodules, it's my build script: https://github.com/afnanenayet/diffsitter/blob/46d4e2ee269b58d118fe1d3543c22c919281820e/build.rs#L58-L82

Good news there is that it's my code so we can do literally anything that Rust allows. We can add a feature flag or some environment variable that disables compiling the grammars directly, but we still need some way for diffsitter to access the grammars so it can compile them, link, and do the codegen step that gives diffsitter access to the FFI functions that initialize the grammars.

So the easiest thing I can think of is that we add some build flag that allows you to specify an arbitrary directory for the grammars, and then the build script could just look there for the grammars.

So it would look something like this:

```sh
export DIFFSITTER_EXTERNAL_GRAMMAR_DIR=$pkgroot/my_grammar_sources # make sure you have every grammar repo in this directory named identically to what I have in my repo
cargo build --features external-grammars --better-build-info # <- might want to use this for distribution
```

Now if we want to deal with dynamic linkage this might be a tad more complex, right now everything is compiled with the cc crate which statically links by default, and I believe that Rust in general is pretty biased towards static linkage, but if you feel strongly about this we can brainstorm a good solution for dynamically loading the tree-sitter libs.

**Comment by lmartinez-mirror** [comment]
Thanks for your response! Unfortunately I'm no Rust developer, but if build flags are the easiest solution, that's fine by me. Each grammar package `tree-sitter-<lang>-git` installs to `/usr/lib/libtree-sitter-<lang>.so`, so at least I can keep them consistent by adding them to `optdepends` or the like.

**Comment by afnanenayet** [comment]
Hey, I created a branch that should address this, could you see if this works for you?

The branch is: `afnan/dynamic-loading`

When building with cargo you can enable loading from dynamic libraries with: `cargo build --features dynamic-grammar-libs`. I also added tests (which also ensure that the libraries are compatible with the binary's version of tree-sitter, which can be done with a similar process: `cargo test --features dynamic-grammar-libs`.

**Comment by lmartinez-mirror** [comment]
Hi, thanks for taking the time to address this!

I've tried to build this a few days ago but still fails, even with the flag. I'll see if I can post a build log over the weekend.

**Comment by afnanenayet** [comment]
oh rip, yeah keep me posted

**Comment by afnanenayet** [comment]
Oh I realized the error is probably because I need to actually supply the file name of the library, I thought that libloading would automatically handle the names on different platforms (ex: `libtree-sitter-c.dylib` on MacOS and `libtree-sitter-c.so` on Linux/BSD. Seems like that's not the case. The documentation also mentions that it's flaky to use filenames and far more reliable to use absolute paths. I'm wondering if it might be easier to have the packager supply a list of tree-sitter libraries at build time or something to the binary, though that's not an ideal solution.

**Comment by afnanenayet** [comment]
I updated the branch so that extensions are applied based on the platform, see if that works?

**Comment by lmartinez-mirror** [comment]
Sorry about the late response. Haven't had time to test this in-depth, but from what I'm seeing the repo still wants to pull the submodules for building. I'll take a closer look later.

**Comment by afnanenayet** [comment]
Ah man my bad, I forgot to update the `build.rs` to only compile the static grammars when the build flag is enabled. That should be fixed now.

**Comment by afnanenayet** [comment]
Also, appreciate your patience :)

**Comment by lmartinez-mirror** [comment]
Alright I've tried to build again and we're getting somewhere. I've attached a build log of an error I get when using the dynamic-libs flag.

[diffsitter-git-0.6.6.r49.g9e0a858-1-x86_64-build.log](https://github.com/afnanenayet/diffsitter/files/7026413/diffsitter-git-0.6.6.r49.g9e0a858-1-x86_64-build.log)

**Comment by afnanenayet** [comment]
Pushed a fix in the branch, I'm surprised this didn't come up when I was testing

**Comment by afnanenayet** [comment]
@lmartinez-mirror @bbigras I've merged support for dynamically linking grammars, I'll do some more testing on my own before cutting a new release, but feel free to test with the `main` branch and let me know if things work as expected.

**Comment by lmartinez-mirror** [comment]
Hi again, finally got around to checking out the latest release. I've tried building this using the dynamic-grammar-libs feature, but I get a build error.

Do you need the submodules present for this? I can give more details if needed.

[diffsitter-0.6.7-1-x86_64-build.log](https://github.com/afnanenayet/diffsitter/files/7664608/diffsitter-0.6.7-1-x86_64-build.log)

**Comment by afnanenayet** [comment]
Nope, you did everything right. Fix incoming in https://github.com/afnanenayet/diffsitter/pull/254

**Comment by afnanenayet** [comment]
@lmartinez-mirror new release should be ok: https://github.com/afnanenayet/diffsitter/releases/tag/v0.6.8

**Comment by lmartinez-mirror** [comment]
This did the trick, thank you! I'll be uploading a compiled 0.6.8 package to the AUR shortly.

**Comment by afnanenayet** [comment]
Sweet, glad it worked
